go/vt/vtgate/vindexes/vschema.go

```diff
@@ -327,18 +353,26 @@ func buildKeyspaceGlobalTables(vschema *VSchema, ksvschema *KeyspaceSchema) {
 			if gt == nil {
 				// Table name is already marked ambiguous, nothing to do.
 				continue
-			} else if t.isReferencedInKeyspace(gt.Keyspace.Name) {
-				// If the stored table refers to this table, store this
-				// table instead.
-				vschema.globalTables[tname] = t
-			} else if gt.isReferencedInKeyspace(t.Keyspace.Name) {
-				// The source of this table is already stored. Do nothing.
-				continue
 			} else {
+				// Special handling for reference tables which specify their source.
+				if t.Type == TypeReference && t.Source != nil {
+					// If the reference table points to the already stored
+					// global table, there is no ambiguity.
+					if t.Source.Qualifier.String() == "" || t.Source.Qualifier.String() == gt.Keyspace.Name {
```

Comment:
nit: you could do `t.Source.Qualifier.IsEmpty()` instead of string comparison here